rfcs/shared/build-system/ssr-testing.md

```diff
@@ -0,0 +1,240 @@
+  // React & dependencies are not bundled intentionally to avoid collisions when they will be required
+  external: ['react', 'scheduler', 'react-dom', '@griffel/react', '@griffel/core'],
+  outfile: './dist/out-node.js',
+  plugins: [TsconfigPathsPlugin({ tsconfig: path.resolve(__dirname, 'tsconfig.base.json') })],
+});
+await esbuild.build({
+  // ESM has different entrypoint as we need to call "ReactDOM.hydrate()" there
+  entryPoints: ['app.tsx'],
+  bundle: true,
+  jsx: 'transform',
+  format: 'iife',
+  target: 'chrome101',
+  outfile: './dist/out-browser.js',
+  plugins: [TsconfigPathsPlugin({ tsconfig: path.resolve(__dirname, 'tsconfig.base.json') })],
+});
+```
+
+</details>
+
+**Note:** ESBuild will use existing path aliases to resolve packages used in stories.
+
+### Prepare assets
+
+On this stage we effectively need to create `index.html` that will be used for browser testing.
+
+We will use APIs for React 16.8/17 as we don't support React 18 yet. Once we will get React 18 in our repo it's reasonable to switch to streaming rendering. The snippet below renders all components and styles for them to static markup and saves as `index.html`.
+
+<details>
+  <summary>`render.js`</summary>
+
+```js
+const App = require('./out-node').App;
+const renderer = createDOMRenderer();
+
+const resultDOM = ReactDOM.renderToString(
+  React.createElement(RendererProvider, { renderer }, React.createElement(App)),
+);
+const resultStyles = ReactDOM.renderToStaticMarkup(renderToStyleElements(renderer));
+
+await fs.promises.writeFile(
+  './index.html',
+  `
+  <html>
+    <head>${resultStyles}</head>
+    <body>
+      <div id="root">${resultDOM}</div>
+      <script src="./out-browser.js"></script>
+    </body>
+  </html>
+  
+  `,
+);
+```
+
+</details>
+
+### Test
+
+This stage performs actual testing.
+
+- Start Chrome (via `puppeteer`)
+  - **Note**: the latest version of Puppeteer will be used to get latest Chrome. Existing version in our repo does not support even `:focus-visible` (that creates false positive errors)
+- Open a page (`index.html` file)
+  - **Note**: Webserver will not be used, a file will be opened via `file://`
+  - A page will have configured interceptors to avoid loading resources (images) from CDN
+- Ensure that there are no errors, otherwise test fails
+  - **Note**: For this reason the test app will run with `process.NODE_ENV = 'development'`
+  - **Note**: _Probably_ there is sense to restrict the test to catch only SSR related errors (in prototype it fails also on missing React keys, for example)
+
+## Summary
+
+- This test is not reusable and will be implemented only for `@fluentui/react-components`
+  - `@fluentui/react` & `@fluentui/react-northstar` has own tests
+- Taking into account the speed, there are reasons to make it scoped
+  - The new package will have `@fluentui/react-components` to trigger the test based on dependencies
+
+## Pros and Cons
+
+- 👍 It's fast (3 seconds to run the whole test)
```

Comment:
Will it be possible to run this test suite locally while working on a component?

Reply:
> Will it be possible to run this test suite locally while working on a component?

Yes, it will be something like `yarn build && yarn test` in that package. So no hot reload or something like, but it should be a fast loop.

> but ATM taking into consideration the speed of running it for everything in one shot is making that a low prio.

Yeah, that's why I ignored scoping part 🐱